### ISSUE TITLE
Fuzzer found bug: Sequence function generates arrays with negative sizes.

### ISSUE THREAD

**Issue by laithsakka** [facebookincubator/velox#6428]
### Bug description

https://app.circleci.com/pipelines/github/facebookincubator/velox/32721/workflows/017a1935-85d0-4c78-b6c4-df12997b7113/jobs/210356/artifacts

### System information

NA

**Comment by laithsakka** [comment]
The issue is in the sequence function
the function 
getStepCount(start, stop, step); returns negative value, this should not be the case the input in that to the function 
getStepCount is 
start= 1420919120
stop = 1476733096
step = 162700

**Comment by laithsakka** [comment]
I assume diffdate is returning -525535 but since 1420919120 is < 1476733096 it should be actually 
a positive number.

**Comment by mbasmanova** [comment]
CC: @kewang1024

**Comment by laithsakka** [comment]
Note that in debug mode this result in VeloxRunTimeError due to debug time checks, but this can result in crashes in opt mode!

**Comment by laithsakka** [comment]
since we are computing the difference i wonder if it would be correct if we rebase the two numbers by substricting the abs diff
say if we have  X, Y . Y>X
we compute diff on 0, Y-X

**Comment by kewang1024** [comment]
> 1420919120

Thanks for the data point! I was able to reproduce the issue locally. 

If I understand correctly, the issue is with the utility function `diffTimestamp`, so we should fix this utility class in general? 
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/DateTimeImpl.h#L135

@gggrace14

**Comment by kewang1024** [comment]
Here is the unit test that can reproduce: https://github.com/facebookincubator/velox/pull/6476

**Comment by kewang1024** [comment]
Reproduce
diffdate(1420919120, 1476733096) returns -525535

because inside `diffTimestamp`
const date::year_month_day toCalDate(toDaysTimepoint);

when toDaysTimepoint is 1476733096, toCalDate is negative
![image](https://github.com/facebookincubator/velox/assets/50636602/7977fcdf-6902-4aec-b12f-a79a207a228a)

I think we can either
1. make changes to `date::year_month_day` to add sanity check, which seems not realistic
2. Or add sanity check before using `date::year_month_day` in DateTimeImpl and throw exceptions when needed

@gggrace14 @laithsakka

**Comment by gggrace14** [comment]
I'm looking at the fix to this issue now, as it involves the date_diff func. Will update later.

**Comment by kewang1024** [comment]
We can’t enforce the sanity check data range in sequence function because it wouldn’t align with java’s behavior.

So we decided that the utility class date_diff has to be fixed to handle larger range than it currently supports, open a new issue to track the general fix on DateTimeImpl::diffDate: https://github.com/facebookincubator/velox/issues/6521

**Comment by gggrace14** [comment]
Original failed fuzzer I see is "sequence"('3892314-06-02'::DATE, '4045127-11-23'::DATE, "c0")

I checked and Java supports the date range and there's no failure. 20230913_061242_20700_ywzub

> The issue is in the sequence function the function getStepCount(start, stop, step); returns negative value, this should not be the case the input in that to the function getStepCount is
> start= 1420919120
> stop = 1476733096
> step = 162700

> I assume diffdate is returning -525535 but since 1420919120 is < 1476733096 it should be actually
a positive number.

If above is the reproduced failure, I think I find the cause, which is traced down to the 3rd party lib, date.h, that we import to velox codebase, https://github.com/facebookincubator/velox/blob/main/velox/external/date/date.h#L397-L399. (Not about int64_t overflow)

diffDate() & diffTimestamp() implementation utilizes the `class year` from date.h, and `class year` uses type `short` for the year value. In the fuzzer case of year 3892314 & 4045127,  16-bit short overflows.

For a quick fix to unblock our GitHub fuzzer test, I suggest we directly change the year class in our codebase to use `int` to represent year value. is my attempt.

On the other hand, still need to figure out the following
1. I see the fuzzer failure is "Reason: (-2 vs. 0) ArrayVectorBase size must be greater than zero. Index: 13." However, 
2. 
3. 
4. 

https://github.com/HowardHinnant/date/blob/cc4685a21e4a4fdae707ad1233c61bbaff241f93/include/date/date.h#L395-L397

**Comment by gggrace14** [comment]
On the other hand, still need to figure out the following

1. I see from the fuzzer log that the failure is
> "Reason: (-2 vs. 0) ArrayVectorBase size must be greater than zero. Index: 13."

However, the failure I got locally by reproducing the case diffDate(1420919120, 1476733096) is

> C++ exception with description "Exception: VeloxUserError
> Error Source: USER
> Error Code: ARITHMETIC_ERROR
> Reason: integer overflow: 18446744073709551614 * 4
> Retriable: False
> Context: sequence(c0, c1, c2)
> Top-Level Context: Same as context.
> Function: checkedMultiply
> File: /Users/gaoge/Program/presto/presto-native-execution/velox/velox/common/base/CheckedArithmetic.h

@laithsakka , can you help confirm that diffDate(1420919120, 1476733096) is the case that fails the Fuzzer? Is above the same thing as you said previously?
> in debug mode this result in VeloxRunTimeError due to debug time checks, but this can result in crashes in opt mode

2. Longer term solution might be to revisit the implementation of diffDate() & diffTimestamp() and use a different 3rd party datetime lib. At the moment I'm not sure which 3rd party lib can be better. date.h was firstly added by Pedro quite a while ago, to be used the same way as Java uses the joda lib. When I implemented diffDate() & diffTimestamp(), I evaluated the existing date.h, boost datetime API and C++ std::chrono, and decided date.h fits better. We can re-evaluate boost, but my memory is building boost needs finer configure in order to have the same integer precision as Java joda for some datetime API.

**Comment by mbasmanova** [comment]
@kewang1024 Ke, what's the status of this issue? Is there anything blocking a fix?

**Comment by Yuhta** [comment]
Is this issue still happening?
https://github.com/facebookincubator/velox/actions/runs/9360225008/job/25766212805?pr=10012

**Comment by mbasmanova** [comment]
Still seeing this https://github.com/facebookincubator/velox/actions/runs/9457326845/job/26052113608?pr=10126

@kewang1024 Ke, would you help take a look?
